### PR TITLE
feat: support multipart/related requests

### LINKED ISSUE #491: feature request: support multipart requests
Currently it does not appear that Gaxios supports the ability to send multipart requests. As maintainers of google-cloud/storage we would like to be able to move away from teeny-request (which does support multipart requests) as well as add the ability to do batch deletes which require multipart requests. It would be nice to add multipart requests to this library.

### PR DESCRIPTION
**Pull request by ddelgrosso1**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/gaxios/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #491 🦕

### CODE REVIEW

File: src/common.ts [hunk 1]
```diff
@@ -135,6 +136,11 @@ export interface GaxiosResponse<T = any> {
   request: GaxiosXMLHttpRequest;
 }
 
+export interface GaxioMultipartOptions {
```
Comment (danielduhh) [inline src/common.ts:139]:
```suggestion
export interface GaxiosMultipartOptions {
```

File: src/common.ts [hunk 2]
```diff
@@ -175,6 +181,7 @@ export interface GaxiosOptions {
    */
   maxRedirects?: number;
   follow?: number;
+  multipart?: GaxioMultipartOptions[];
```
Comment (danielduhh) [inline src/common.ts:184]:
```suggestion
  multipart?: GaxiosMultipartOptions[];
```

File: src/gaxios.ts [hunk 3]
```diff
@@ -294,6 +296,19 @@ export class Gaxios {
       } else {
         opts.body = opts.data;
       }
+    } else if (opts.multipart && opts.multipart.length > 0) {
+      // note: once the minimum version reaches Node 16,
```
Comment (danielduhh) [inline src/gaxios.ts:300]: Can you create a ticket for this?

File: src/gaxios.ts [hunk 4]
```diff
@@ -416,4 +431,30 @@ export class Gaxios {
...
     }
   }
+
+  /**
+   * Creates an async generator that yields the pieces of a multipart/related request body.
+   *
+   * @param {GaxioMultipartOptions[]} multipartOptions the pieces to turn into a multipart/related body.
+   * @param {string} boundary the boundary string to be placed between each part.
+   */
+  private async *getMultipartRequest(
```
Comment (danielduhh) [inline src/gaxios.ts:441]: Is this based on a specific spec, or is this just what we need in Storage?

**Review by danielduhh — COMMENTED** [review]
Small nit and a question

File: src/gaxios.ts (hunk 3 above)
Comment (ddelgrosso1) [reply src/gaxios.ts:300]: Done. b/332588446

File: src/gaxios.ts (hunk 4 above)
Comment (ddelgrosso1) [reply src/gaxios.ts:441]: This is based on https://www.ietf.org/rfc/rfc2387.txt with a slight caveat. We are not currently supporting multipart/related within a part (i.e. recursive support).

**Review by danielbankhead — APPROVED** [review]
Looks great! Thanks for adding

File: src/gaxios.ts (hunk 4 above)
Comment (danielduhh) [reply src/gaxios.ts:441]: sg. Do you mind adding a note in the comments?

**Review by danielduhh — APPROVED** [review]

File: src/gaxios.ts (hunk 4 above)
Comment (ddelgrosso1) [reply src/gaxios.ts:443]: Done

**Pull request merged**